examples/seismic/viscoacoustic/operators.py

```diff
@@ -63,9 +62,87 @@ def blanch_symes(model, geometry, v, p, **kwargs):
     return [u_v, u_r, u_p]
 
 
-def ren(model, geometry, v, p, **kwargs):
+def sls_2nd_order(model, geometry, v, p, **kwargs):
+    """
+    Implementation of the 2nd order viscoacoustic wave-equation from Bai (2014).
+
+    https://library.seg.org/doi/10.1190/geo2013-0030.1
+
+    Parameters
+    ----------
+    v : VectorTimeFunction
+        Particle velocity.
+    p : TimeFunction
+        Pressure field.
+    """
+    forward = kwargs.get('forward', True)
+    space_order = p.space_order
+    s = model.grid.stepping_dim.spacing
+    b = model.b
+    vp = model.vp
+    damp = model.damp
+    qp = model.qp
+    f0 = geometry._f0
+
+    # The stress relaxation parameter
+    t_s = (sp.sqrt(1.+1./qp**2)-1./qp)/f0
+
+    # The strain relaxation parameter
+    t_ep = 1./(f0**2*t_s)
+
+    # The relaxation time
+    tt = (t_ep/t_s)-1.
+
+    # Density
+    rho = 1. / b
+
+    r = TimeFunction(name="r", grid=model.grid, time_order=2, space_order=space_order,
+                     staggered=NODE)
+
+    if forward:
+
+        u_v = Eq(v, b * grad(p))
+
+        pde_r = r + s * (tt / t_s) * rho * div(v) - s * (1. / t_s) * r
+        u_r = Eq(r.forward, damp * pde_r)
+
+        pde_p = 2. * p - damp * p.backward + s * s * vp * vp * (1. + tt) * rho * \
+            div(v) - s * s * vp * vp * r.forward
+        u_p = Eq(p.forward, damp * pde_p)
+
+        return [u_v, u_r, u_p]
+
+    else:
+
+        # Auxiliary functions
+        w = VectorTimeFunction(name="w", grid=model.grid, time_order=2,
+                               space_order=space_order)
+        h = TimeFunction(name="h", grid=model.grid, time_order=2, space_order=space_order,
+                         staggered=NODE)
+        g = TimeFunction(name="g", grid=model.grid, time_order=2, space_order=space_order,
+                         staggered=NODE)
+
+        pde_r = r + s * (tt / t_s) * p - s * (1. / t_s) * r
+        u_r = Eq(r.backward, damp * pde_r)
+
+        u_h = Eq(h, (1. + tt) * rho * p)
+
+        u_v = Eq(v, b * grad(h))
+
+        u_g = Eq(g, rho * r.backward)
+
+        u_w = Eq(w, b * grad(g))
```

Comment:
Are these temps really necessary? Is there an issue if you just put
```
 pde_p = 2. * p - damp * p.forward + s * s * vp * vp * div(b * grad((1. + tt) * rho * p))) - \
            s * s * vp * vp * div(b * grad(rho * r.backward))

```

You can still split the expression in a few lines if you want to avoid one long expression but the extra function/vector shulnd't be necessary.

Reply:
We appreciate your review @mloubout .
These temps were used to avoid the known and described issue about multiple functions on the argument of `grad` operator `Ex: grad((1. + tt) * rho * p)` . 
So we compute each step of the parameters before passing them as arguments.

Reply:
Is it still an issue? I think `grad` is used some places on expressions and should support it. If it fails, could you open an issue with an MFE on it?

Reply:
Yes, until the day I did the PR, this is still an issue.
You fixed it for the `div`, but not for the `grad` yet.
Yes, I can open an issue. I will do this as soon as possible.

Reply:
```
In [16]: div(u * grad((1+t) * u * v))                                                                                                                                                
Out[16]: Derivative(u(t, x, y)*Derivative(3*u(t, x, y)*v(x, y), x), x) + Derivative(u(t, x, y)*Derivative(3*u(t, x, y)*v(x, y), y), y)

```

Works fine for expressions this is weird.

Reply:
I see. But when the stencil is passed to the operator the results produce NAN. Anyway, I will check on Monday with the most updated version of Devito, and if confirmed I will open an issue. Otherwise, I will remove the tmps.

Reply:
>passed to the operator the results produce NAN.

Now that's something I can get behind, may need some parametrizable grad/div  to keep it nice and stable (i.e like how you can do `u.dx(x0=x + dx)` but for grad/div. I'll see if I can find time to play around with your PR and file a patch for it but may take a little bit so can leave the implementation like that.

Reply:
Got it. I will keep implementing it that way.
Thanks for the clarification.
On Monday I will make the suggested corrections.